PhysicsTools/TensorFlow/src/NTSession.cc

```diff
@@ -0,0 +1,1442 @@
+// class instances and edgeset adds a member data if NDEBUG is set
+
+/*
+This file is an adaptation of the original direct_session.cc file located at
+https://github.com/tensorflow/tensorflow/blob/v1.3.0/tensorflow/core/common_runtime/direct_session.cc
+to meet the demands of the software environment developed and used by the CMS collaboration.
+
+Changes with respect to the original code are documented in the NTSession.h header file.
+*/
+
+#if !defined(NDEBUG)
+#define NDEBUG 1
+#endif
+
+#include "NTSession.h"
+
+#include <atomic>
+#include <string>
+#include <vector>
+
+#include "tensorflow/core/common_runtime/constant_folding.h"
+#include "tensorflow/core/common_runtime/debugger_state_interface.h"
+#include "tensorflow/core/common_runtime/device_factory.h"
+#include "tensorflow/core/common_runtime/executor.h"
+#include "tensorflow/core/common_runtime/function.h"
+#include "tensorflow/core/common_runtime/graph_optimizer.h"
+#include "tensorflow/core/common_runtime/memory_types.h"
+#include "tensorflow/core/common_runtime/optimization_registry.h"
+#include "tensorflow/core/common_runtime/simple_placer.h"
+#include "tensorflow/core/common_runtime/step_stats_collector.h"
+#include "tensorflow/core/framework/function.h"
+#include "tensorflow/core/framework/graph.pb_text.h"
+#include "tensorflow/core/framework/graph.pb.h"
+#include "tensorflow/core/framework/graph_def_util.h"
+#include "tensorflow/core/framework/log_memory.h"
+#include "tensorflow/core/framework/node_def.pb.h"
+#include "tensorflow/core/framework/tensor.h"
+#include "tensorflow/core/framework/versions.pb.h"
+#include "tensorflow/core/graph/algorithm.h"
+#include "tensorflow/core/graph/graph.h"
+#include "tensorflow/core/graph/graph_constructor.h"
+#include "tensorflow/core/graph/graph_partition.h"
+#include "tensorflow/core/graph/subgraph.h"
+#include "tensorflow/core/graph/tensor_id.h"
+#include "tensorflow/core/lib/core/errors.h"
+#include "tensorflow/core/lib/core/notification.h"
+#include "tensorflow/core/lib/core/refcount.h"
+#include "tensorflow/core/lib/core/status.h"
+#include "tensorflow/core/lib/gtl/array_slice.h"
+#include "tensorflow/core/lib/gtl/stl_util.h"
+#include "tensorflow/core/lib/monitoring/counter.h"
+#include "tensorflow/core/lib/strings/numbers.h"
+#include "tensorflow/core/lib/strings/str_util.h"
+#include "tensorflow/core/lib/strings/strcat.h"
+#include "tensorflow/core/platform/cpu_info.h"
+#include "tensorflow/core/platform/logging.h"
+#include "tensorflow/core/platform/mutex.h"
+#include "tensorflow/core/platform/types.h"
+#include "tensorflow/core/util/device_name_utils.h"
+#include "tensorflow/core/util/env_var.h"
+
+#if GOOGLE_CUDA
+#include "tensorflow/core/common_runtime/gpu/gpu_tracer.h"
+#endif  // GOOGLE_CUDA
+
+namespace tensorflow {
+
+namespace {
+
+auto* nothreads_session_runs = monitoring::Counter<0>::New(
+    "/tensorflow/core/nothreads_session_runs",
+    "The number of times NTSession::Run() has been called.");
```

Comment:
the static analyzer marks this as a thread safety issue
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-19893/25360/llvm-analysis/report-9c4c54.html#EndPath

@Dr15Jones 
is this actually safe in this context?
If it is safe, a ```CMS_THREAD_SAFE ``` should be added in front of the variable definition.
This will need ```#include "FWCore/Utilities/interface/thread_safety_macros.h"```

Reply:
@slava77 we could add the declaration once @Dr15Jones confirms the safeness

Reply:
@Dr15Jones 
please take a look
Thank you.

Reply:
The Counters do appear safe: see
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/lib/monitoring/counter.h